File: app/libs/programming/Attribute.js
class Attribute {

    constructor(name){
        this.name = name
        this.type = "String"
        this.defaultValue = ""
        this.isDefault = false
        this.isArray = false
        this.isRequire = false
        this.isLowercase = false
        this.isTrim = false
        this.max = 0
        this.min = 0
        this.isMax = false
        this.isMin = false
        this.isUnique = false
        this.isEncrypt = false
        this.isGet = true
        this.isUpdate = true
        this.isAuth = false
        this.isId = false
        this.isObject = false
    }

    setTypeString(){
        this.type = "String"
        return this
    }

    setTypeId(ref){
        this.type = `mongoose.Schema.Types.ObjectId , ref: '${ref}'`
        return this
    }

    setTypeDate(){
        this.type = "Date"
        return this
    }

    setTypeNumber(){
        this.type ="Number"
        return this
    }

    setTypeBuffer(){
        this.type = "Buffer"
        return this
    }

    setTypeMixed(){
        this.type = "mongoose.Schema.Types.Mixed"
        return this
    }

    setTypeBoolean(){
        this.type = "Boolean"
        return this
    }

    setTypeMap(){
        this.type = "Map"
        return this
    }

    setTypeDecimal128(){
        this.type = "mongoose.Schema.Types.Decimal128"
        return this
    }

    setTypeObject(objs){
        
        let obj = ""

        objs.map((o) => {
            obj += `${o.name} : ${o.generate()},\n`
        })

        this.type = obj
        this.isObject = true
        return this
    }

    setTypeToArray(){
        this.isArray = true
        return this
    }

    require(){
        this.isRequire = true
        return this
    }

    default(defaultValue){
        this.isDefault = true
        this.defaultValue = defaultValue
        return this
    }

    lowercase(){
        this.isLowercase = true
        return this
    }

    trim(){
        this.isTrim = true
        return this
    }

    setMax(value){
        this.max = value
        this.isMax = true
        return this
    }

    setMin(value){
        this.min = value
        this.isMin = true
        return this
    }

    unique(){
        this.isUnique = true
        return this
    }

    encrypt(){
        this.isEncrypt = true
        return this
    }

    auth(){
        this.isAuth = true
        return this
    }

    id(){
        this.isId = true
        return this
    }

    disableGet(){
        this.isGet = false
        return this
    }

    disableUpdate(){
        this.isUpdate = false
        return this
    }

    generate(){

        let code = this.type

        if (this.isRequire){
            code += ", required : true"
        }

        if (this.isDefault){
            code += `, default : ${this.defaultValue}` 
        }

        if (this.isLowercase && this.type === "String"){
            code += ", lowercase : true"
        }

        if (this.isTrim && this.type === "String"){
            code += ", trim : true"
        }

        if (this.isMax && (this.type === "Number" || this.type === "mongoose.Schema.Types.Decimal128")){
            code += `, max : ${this.max}`
        }

        if (this.isMin && (this.type === "Number" || this.type === "mongoose.Schema.Types.Decimal128")){
            code += `, min : ${this.min}`
        }

        if (this.isObject){
            code = `{ \n${code} }`
        }else{
            code = `{ type: ${code} }`
        }

        if (this.isArray){
            code = `[${code}]`
        }

        return code
    }

}

module.exports = Attribute

File: app/libs/programming/attribute.py
class Attribute:

	def __init__(self , name):
		self.name = name
		self.type = "String"
		self.defaultValue = ""
		self.isDefault = False
		self.array = False
		self.isRequire = False
		self.isLowercase = False
		self.isTrim = False
		self.max = 0
		self.min = 0
		self.isMax = False
		self.isMin = False
		self.isUnique = False
		self.isEncrypt = False
		self.isGet = True
		self.isUpdate = True
		self.isAuth = False
		self.isId = False

	def setTypeString(self):
		self.type = "String"
		return self

	def setTypeID(self , ref):
		self.type = "mongoose.Schema.Types.ObjectId , ref: '{0}'".format(ref)
		return self

	def setTypeDate(self):
		self.type = "Date"
		return self

	def setTypeNumber(self):
		self.type = "Number"
		return self

	def setTypeBuffer(self):
		self.type = "Buffer"
		return self

	def setTypeMixed(self):
		self.type = "mongoose.Schema.Types.Mixed"
		return self

	def setTypeBoolean(self):
		self.type = "Boolean"
		return self

	def setTypeMap(self):
		self.type = "Map"
		return self

	def setTypeDecimal128(self):
		self.type = "mongoose.Schema.Types.Decimal128"
		return self

	def setTypeObject(self , objs):
		obj = ""
		for o in objs:
			obj += "{ "+ o.generate() +" },"
		self.type = obj
		return self

	def setTypeToArray(self):
		self.array = True
		return self

	def clearType(self):
		self.type = "String"
		self.defaultValue = ""
		self.isDefault = False
		self.array = False
		self.isRequire = False
		self.isLowercase = False
		self.isTrim = False
		self.max = 0
		self.min = 0
		self.isMax = False
		self.isMin = False
		self.isUnique = False
		self.isEncrypt = False
		self.isGet = True
		self.isUpdate = True
		self.isAuth = False
		self.isId = False
		return self

	def require(self):
		self.isRequire = True
		return self

	def default(self , value):
		self.defaultValue = str(value)
		self.isDefault = True
		return self

	def lowercase(self):
		self.isLowercase = True
		return self

	def trim(self):
		self.isTrim = True
		return self

	def setMax(self , value):
		self.max = value
		self.isMax = True
		return self

	def setMin(self , value):
		self.min = value
		self.isMin = True
		return self

	def unique(self):
		self.isUnique = True
		return self

	def encrypt(self):
		self.isEncrypt = True
		return self

	def auth(self):
		self.isAuth = True
		return self

	def id(self):
		self.isId = True
		return self

	def getIsAuth(self):
		return self.isAuth

	def disableGet(self):
		self.isGet = False
		return self

	def disableUpdate(self):
		self.isUpdate = False
		return self

	def getIsGet(self):
		return self.isGet

	def getIsUpdate(self):
		return self.isUpdate

	def getIsEncrypt(self):
		return self.isEncrypt

	def getIsUnique(self):
		return self.isUnique

	def getName(self):
		return self.name

	def getIsId(self):
		return self.isId

	def generate(self):
		code = self.type

		if self.isRequire:
			code += ", required : true"

		if self.isDefault:
			code += " , default :" + self.defaultValue

		if self.isLowercase:
			code += ", lowercase : true"

		if self.isTrim:
			code += ", trim : true"

		if self.isMax:
			code += ", max : " + str(self.max)

		if self.isMin:
			code += ", min :" + str(self.min)

		code = "{ type:" + code + "}"

		if self.array:
			code = "[" + code + "]"

		return code
